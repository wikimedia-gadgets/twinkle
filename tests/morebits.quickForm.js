/** Simple helper to render a quickform element */
function renderElement(data) {
	return new Morebits.quickForm.element(data).render();
}

describe("quickform", () => {
	let inputConfig = {
		type: "input",
		label: "Label",
		name: "inputname",
		value: "prefilled value",
	};
	let selectConfig = {
		type: "select",
		label: "Select label",
		name: "selectname",
		list: [
			{ label: "select label 1", value: "selectvalue1" },
			{ label: "select label 2", value: "selectvalue2" },
		],
	};
	let checkboxListConfig = {
		type: "checkbox",
		name: "checkboxlist1",
		list: [
			{ label: "checkbox 1", value: "checkval1" },
			{ label: "checkbox 2", value: "checkval2", checked: true },
		],
	};
	let checkboxesConfig = {
		type: "checkbox",
		list: [
			{
				label: "checkbox 1",
				value: "checkval1",
				name: "checkname1",
				checked: true,
			},
			{ label: "checkbox 2", value: "checkval2", name: "checkname2" },
		],
	};
	let radiosConfig = {
		type: "radio",
		name: "radiolist1",
		list: [
			{
				label: "radio 1",
				value: "radioval1",
				name: "radioname1",
			},
			{
				label: "radio 2",
				value: "radioval2",
				name: "radioname2",
				checked: true,
			},
		],
	};
	let textareaConfig = {
		type: "textarea",
		name: "textareaname1",
		value: "textarea prefilled value",
		cols: 4,
	};
	let submitConfig = {
		type: "submit",
	};
	let buttonConfig = {
		type: "button",
		label: "button label",
	};

	let getRenderedForm = () => {
		let form = new Morebits.quickForm(function () {});
		form.append(inputConfig);
		form.append(checkboxesConfig);
		form.append(checkboxListConfig);
		form.append(radiosConfig);
		form.append(textareaConfig);
		form.append(submitConfig);
		form.append(buttonConfig);
		return form.render();
	}

	test("input element", () => {
		expect(renderElement(inputConfig)).toMatchSnapshot();
	});

	test("select element", () => {
		expect(renderElement(selectConfig)).toMatchSnapshot();
	});

	test("checkbox elements", () => {
		expect(renderElement(checkboxListConfig)).toMatchSnapshot();
	});

	test("checkbox elements (with data-single attribute)", () => {
		expect(renderElement(checkboxesConfig)).toMatchSnapshot();
	});

	test("radio elements", () => {
		expect(renderElement(radiosConfig)).toMatchSnapshot();
	});

	test("textarea element", () => {
		expect(renderElement(textareaConfig)).toMatchSnapshot();
	});

	test("submit", () => {
		expect(renderElement(submitConfig)).toMatchSnapshot();
	});

	test("getInputData", () => {
		expect(Morebits.quickForm.getInputData(getRenderedForm())).toMatchSnapshot();
	});

	test("getElements", () => {
		let form = getRenderedForm();
		expect(Morebits.quickForm.getElements(form, 'checkboxlist1')).toMatchSnapshot();
	});

});
