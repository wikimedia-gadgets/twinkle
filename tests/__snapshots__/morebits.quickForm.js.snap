// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quickform checkbox elements (with data-single attribute) 1`] = `
<div
  id="node_5"
>
  <div>
    <input
      checked="checked"
      data-single="data-single"
      id="node_5_0"
      name="checkname1"
      type="checkbox"
      value="checkval1"
    />
    <label
      for="node_5_0"
    >
      checkbox 1
    </label>
  </div>
  <div>
    <input
      data-single="data-single"
      id="node_5_1"
      name="checkname2"
      type="checkbox"
      value="checkval2"
    />
    <label
      for="node_5_1"
    >
      checkbox 2
    </label>
  </div>
</div>
`;

exports[`quickform checkbox elements 1`] = `
<div
  id="node_4"
>
  <div>
    <input
      id="node_4_0"
      name="checkboxlist1"
      type="checkbox"
      value="checkval1"
    />
    <label
      for="node_4_0"
    >
      checkbox 1
    </label>
  </div>
  <div>
    <input
      checked="checked"
      id="node_4_1"
      name="checkboxlist1"
      type="checkbox"
      value="checkval2"
    />
    <label
      for="node_4_1"
    >
      checkbox 2
    </label>
  </div>
</div>
`;

exports[`quickform getElements 1`] = `
Array [
  <input
    id="node_20_0"
    name="checkboxlist1"
    type="checkbox"
    value="checkval1"
  />,
  <input
    checked="checked"
    id="node_20_1"
    name="checkboxlist1"
    type="checkbox"
    value="checkval2"
  />,
]
`;

exports[`quickform getInputData 1`] = `
Object {
  "checkboxlist1": Array [
    "checkval2",
  ],
  "checkname1": true,
  "checkname2": false,
  "inputname": "prefilled value",
  "radioname2": "radioval2",
  "textareaname1": "textarea prefilled value",
}
`;

exports[`quickform input element 1`] = `
<div
  id="div_node_0"
>
  <label
    for="node_0"
    style="margin-right: 3px;"
  >
    Label
  </label>
  <input
    id="node_0"
    name="inputname"
    type="text"
    value="prefilled value"
  />
</div>
`;

exports[`quickform radio elements 1`] = `
<div
  id="node_6"
>
  <div>
    <input
      data-single="data-single"
      id="node_6_0"
      name="radioname1"
      type="radio"
      value="radioval1"
    />
    <label
      for="node_6_0"
    >
      radio 1
    </label>
  </div>
  <div>
    <input
      checked="checked"
      data-single="data-single"
      id="node_6_1"
      name="radioname2"
      type="radio"
      value="radioval2"
    />
    <label
      for="node_6_1"
    >
      radio 2
    </label>
  </div>
</div>
`;

exports[`quickform select element 1`] = `
<div
  id="div_node_1"
>
  <label
    for="node_1"
    style="margin-right: 3px;"
  >
    Select label
  </label>
  <select
    id="node_1"
    name="selectname"
  >
    <option
      id="node_2"
      label="select label 1"
      value="selectvalue1"
    >
      select label 1
    </option>
    <option
      id="node_3"
      label="select label 2"
      value="selectvalue2"
    >
      select label 2
    </option>
  </select>
</div>
`;

exports[`quickform submit 1`] = `
<span>
  <input
    id="node_8"
    name="submit"
    type="submit"
  />
</span>
`;

exports[`quickform textarea element 1`] = `
<div
  id="div_node_7"
>
  <textarea
    cols="4"
    id="node_7"
    name="textareaname1"
  />
</div>
`;
